### ISSUE TITLE
local efficiency bug?

### ISSUE THREAD

**Issue by vlado-s** [networkx/networkx#2710]
Hi all,
I wrote my own routines for calculation of efficiency in weighted graphs. While testing them for un-weigted graph (all w=1) I discovered a difference in the values for local efficiency (global if fine). I believe there is a logical bug in the nx.local_efficiency(G) source which is:

return sum(global_efficiency(nx.ego_graph(G, v)) for v in G) / len(G)

however should be:

return sum(global_efficiency(nx.ego_graph(G, v, center=False)) for v in G) / len(G)

I conclude so based on the original paper by Latora & Marchiori (DOI: 10.1103/PhysRevLett.87.198701), where they state (page 2): "Since _E_ is also defined for a disconnected graph we can characterize the local properties of **G** by evaluating for each vertex _i_ the efficiency of **G**_i_ , the subgraph of the neighbors of _i_." Thus _i_ should not be in the ego_graph(G). How do you see it, am I right? Please let me know. Thank's a lot!
Cheers,
Vlado Sladek

**Comment by dschult** [comment]
I interpret the paper as you do. It also matches how Wikipedia defines local efficiency.
Would you like to create a pull request with this change and a test that it is correct?
Or would you prefer that I do that.

**Comment by dschult** [comment]
Take a look at #2233 where this (mis)behavior was implemented apparently to avoid an error raised when the center node is not present.

**Comment by vlado-s** [comment]
Hi, thanks' for your response.
What exactly does the creation of a pull include? What should the testing include? I tested it on one example where I coded it both from scratch & with the modification as written above, and it is then identical to all digits that are displayed with print. Sorry for my ignorance, but are you from the NetworkX developers - i.e. is it easier for you to make the pull? In principle all that needed to be done is what I wrote in the first message, so no big coding is involved. Obviously this works for unweighted graphs (which is what's in NetworkX 2 anyway).

As to the issue https://github.com/networkx/networkx/issues/2233 : seems like the creation with G.subgraph(G[v]) is somewhat cumbersome. If _v_ is a list, then it would be sufficient to write G.subgraph(v). That is not explained there. But later on _v_ is used as a node index in the nx.ego_graph(G, v) command. I used the G.subgraph(list_of_v_neighbours) as well as ego_graph(G, v, center=False) and both yield identical results. The former was my initial idea, just because I had the neighbour list stored from previous steps anyway. Of course _list_of_v_neighbours_ did not include _v_.

**Comment by dschult** [comment]
It will almost certainly be easier for me to create a pull request due to familiarity with github and the codebase/tests etc. But I didn't want to do that if you were already working on it.

```G[v]``` is an easy way to get all neighbors of v. It is actually a dict keyed by neighbors of v with values being the edge attribute dicts, but iterating over ```G[v]``` provides the neighbors of v in the graph. In the end we need ```G.subgraph(neighbors_of_v)``` and ```G[v]``` is an easy way to get the neighbors.

I'll put together a pull request soon.

**Comment by vlado-s** [comment]
OK, sure, no problem, you can make the pull. The main goal is to come to the correct interpretation (which we, I believe, did).
Didn't know about the G[v] syntax... thanks'.

**Comment by vlado-s** [comment]
BTW, sorry that I did not recognize you :) Now I feel stupid...!

**Comment by bakerwho** [comment]
@dschult if this might be a good first issue for someone reasonably experienced with nx and github, but who has never contributed before - I'd appreciate your mentorship.

**Comment by dschult** [comment]
This Issue requires
1) figuring out if we are correct that the local_efficiency should not include the node in question in its neighborhood.  The comments here suggest it should not.
2) If it should not include the node in question then we need to look at #2233 to figure out why that error was occurring and provide a fix (if it hasn't been fixed already). That error involves a report of ```PathNotFound``` when the graph is not connected. I believe the fix for that spurious error is that the global efficiency computation should contribute zero for that pair of nodes (infinite distance provides zero efficiency). But that needs to be determined. It is also possible that a fix has already been put in place for the case of infinite distance between nodes. I vaguely recall that being discussed, but it may have been for a different routine.
3) We should provide some good citations to the info that provides the answers to these questions.
4) Finally, we get down to changing the code -- which I believe should be simple to change.

So, while the code changes here will be fairly simple, the present hurdle is figuring out what actually needs to be done.  If this is what you are looking for then take it on!

Other tips are to read the Wiki page for style features and to look at other modules for examples of how to put together a new module. 

:)

**Comment by vlado-s** [comment]
Hi,
actually, the issue reported in https://github.com/networkx/networkx/issues/2233 may have been solved by someone because now the code for the "pair" efficiency reads (found in the official NX documents https://networkx.github.io/documentation/stable/_modules/networkx/algorithms/efficiency.html#global_efficiency):

def efficiency(G, u, v):
try:
    eff = 1 / nx.shortest_path_length(G, u, v)
except NetworkXNoPath:
    eff = 0
return eff

so the exception that there is no path between two nodes is handled. I just tested it on the exact example as in https://github.com/networkx/networkx/issues/2233 and the local efficiency is calculated by the nx.local_efficiency(G) function without problem (it is 0.9583333333333334). Then I recalculated it by my own routine (keeping the central node in the local sub-graph (i.e. the ego_graph)) and it gave the same result. 

When we consider that the central node is omitted, which can be achieved as I wrote in the original message by doing:
loc_eff_u2=sum(nx.global_efficiency(nx.ego_graph(G, v, center=False)) for v in G) / len(G)
then the value is 0.5833333333333333.

As to **point 2**.; the dependencies are as follows: function local_efficiency calls global_efficiency and that calls efficiency. By omitting the central node in the local ego-graph we can in principle create two "difficult" cases. Firstly, the local ego-graph may contain node pairs that are not connected - this seems to handled well by the exception in the "pair" efficiency function. Secondly, if the central node was already disconnected in the graph G, then its ego-graph is an empty graph - this seems to be handled by the safety condition in global_efficiency:

def global_efficiency(G):
n = len(G)
denom = n * (n - 1)
if denom != 0:
    g_eff = sum(efficiency(G, u, v) for u, v in permutations(G, 2)) / denom
else:
    g_eff = 0

This I tested this too (by removing the edge 1,4 from the example as given in https://github.com/networkx/networkx/issues/2233). The local efficiency is then 0.75.

As to **point 1** and **point 3**. We already agreed that the interpretation of the Latora & Marchiori (DOI: 10.1103/PhysRevLett.87.198701) paper is such that the centre node should be omitted from the local ego-graph. So I think that the reference to this paper is actually satisfactory... I let that to be concluded by you, of course. The eq.1 in the paper clearly, in my view, says that pair contribution is proportional to 1/d_ij, which is zero for disconnected nodes (and is so coded in the global_efficiency function).

Thank you for the devotion to this issue. So it seems as the only thing to be done is to change the function local_efficiency from
return sum(global_efficiency(nx.ego_graph(G, v)) for v in G) / len(G)
to 
return sum(global_efficiency(nx.ego_graph(G, v, center=False)) for v in G) / len(G)

Best,
Vlado

**Comment by bakerwho** [comment]
I have investigated this and I think it looks fairly straightforward. Thanks @dschult for referencing #2233 and @vlado-s for your comprehensive testing of the nature of the problem. In fact, the solution is simply the change @vlado-s proposed.

1. The subgraph used to compute local_efficiency should **not** include the node for which it is being computed. Vlado's quoting of Latora et al confirms this.

2. The error in #2233 was a result of incorrect use of nx.Graph.subgraph(). It seems that the previous implementation was always calling subgraph() on just 1 node of the Graph - leading to a subgraph of just that node, and no edges. This is because subgraph needs to be called on a *subset* of nodes.

Thus, for any graph G:
```python
for node in G:
   G.subgraph(G.neighbors(node)) == nx.ego_graph(G, node, center = False) # True for all nodes
   G.subgraph(node) == nx.ego_graph(G, node, center = False) # as previously implemented in 2233, incorrect
```

3. Vlado has cited the exact lines that are relevant. Should this be explicitly referenced in the PR?

4. I can make the change. As you said, it is exceedingly straightforward.

**Comment by dschult** [comment]
Fixed by #2843
